pkg/chains/formats/slsa/internal/material/material.go

```diff
@@ -77,38 +75,28 @@ func PipelineMaterials(ctx context.Context, pro *objects.PipelineRunObject) ([]c
 		}
 		mats = append(mats, m)
 	}
-	pSpec := pro.Status.PipelineSpec
-	if pSpec != nil {
-		pipelineTasks := append(pSpec.Tasks, pSpec.Finally...)
-		for _, t := range pipelineTasks {
-			tr := pro.GetTaskRunFromTask(t.Name)
-			// Ignore Tasks that did not execute during the PipelineRun.
-			if tr == nil || tr.Status.CompletionTime == nil {
-				logger.Infof("taskrun status not found for task %s", t.Name)
-				continue
-			}
```

Comment:
I think the previous way is safer to do because it lets you catch cases where the Tasks failed to load - the PipelineRun remains the source of truth of what TaskRuns that should be present.

We defer loading Tasks so we don't frontload fetches if there's no work to do, but this also means that a PipelineRunObject may not have a fully produced result for GetTaskRuns() depending on how you call it, and a caller wouldn't know whether this is complete or not.

e.g. how would you detect something like:

```go
o := NewPipelineRunObject(pr)
o.GetTaskRuns() // This is empty
```

We may want to keep things as they were before - IIUC, this already has the same data?

Reply:
I think the old code has same results as the new change, but the new code has cleaner logic.

The formatting code will not be invoked until [controller logic](https://github.com/tektoncd/chains/blob/ddf41fbba572a313f9f21c3549c0fa0e04f91195/pkg/reconciler/pipelinerun/pipelinerun.go#L95-L119) observes all runs finished successfully, and fetches child taskruns and registers it under the pipelinerun.

Reply:
friendly ping for any more thoughts? @wlynch Thanks!

Reply:
[Controller logic](https://github.com/tektoncd/chains/blob/ddf41fbba572a313f9f21c3549c0fa0e04f91195/pkg/reconciler/pipelinerun/pipelinerun.go#L106-L119) already make sure that tr and completion time are not nil before registering the trs to the pipeline.

Reply:
Ah ok! I missed it. So we had this redundant check that should have been removed in the first place?

Reply:
In that case, feel free to resolve my other comments without any action.

Reply:
> Ah ok! I missed it. So we had this redundant check that should have been removed in the first place?

Yeah, exactly!

Reply:
While the code results in the same behavior for Chains overall, it's brittle in the sense that it requires callers to need to pre-populate the TaskRuns, or else they can get wrong data.

My concern is this isn't obvious from an outside point-of-view, and it makes it easy to mistakenly fetch partial/no data without realizing - I'd want something like this to error out if the Object wasn't in the correct state.

I think there's a few options to address this:

1. Revert to the old behavior, where we use the PR status as the source of truth (because then it's obvious there's something wrong if there's a mismatch).
2. Wrap Task fetching in a [sync.Once](https://pkg.go.dev/sync#Once) and lazily fetch it whenever a client tries to read the data, which will guarantee that the data is always fetched (or else error).

Reply:
I feel like there might be some misalignment on the old behaviour.

The old behaviour doesn't not fetch TaskRuns as needed using PR status as source of truth. Instead it also assumes the caller pre-populates the TaskRuns, which is reflected on line 84 in which GetTaskRunFromTask just searches the TaskRun object based on the name from the pre-populated `PipelineRunObject.taskruns` list.


```go
   tr := pro.GetTaskRunFromTask(t.Name)
```

Reply:
Spoke with @wlynch offline. TL;DR: we will keep the existing behaviour and close this PR.

The good thing about the old behaviour is that it can cross check what's reported in the PR status and what's populated to the `PipelineRunObject.taskruns` list, and reports/logs if TaskRuns are not populated, which is important when others use Chains library to reproduce the provenance generation.